src/xmpp/xmpp-im/xmpp_carbons.cpp

```diff
@@ -0,0 +1,222 @@
+/*
+ * xmpp_carbons.cpp - Message Carbons (XEP-0280)
+ * Copyright (C) 2019  Aleksey Andreev
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+ *
+ */
+
+#include <QDomElement>
+
+#include "xmpp_carbons.h"
+#include "xmpp_xmlcommon.h"
+#include "xmpp_message.h"
+#include "xmpp_forwarding.h"
+#include "xmpp_task.h"
+#include "xmpp_tasks.h"
+
+namespace XMPP
+{
+
+static const QString xmlns_carbons(QStringLiteral("urn:xmpp:carbons:2"));
+
+
+class CarbonsSubscriber : public JT_PushMessage::Subscriber
+{
+public:
+        bool xmlEvent(const QDomElement &root, QDomElement &e, Client *client, int userData, bool nested) override;
+        bool messageEvent(Message &msg, int userData, bool nested) override;
+
+private:
+        Forwarding frw;
+};
+
+class JT_MessageCarbons : public Task
+{
+    Q_OBJECT
+
+public:
+    JT_MessageCarbons(Task *parent);
+
+    void enable();
+    void disable();
+
+    void onGo() override;
+    bool take(const QDomElement &e) override;
+
+private:
+    QDomElement iq;
+};
+
+//----------------------------------------------------------------------------
+// JT_MessageCarbons
+//----------------------------------------------------------------------------
+JT_MessageCarbons::JT_MessageCarbons(Task *parent)
+    : Task(parent)
+{
+}
+
+void JT_MessageCarbons::enable()
+{
+    iq = createIQ(doc(), QLatin1String("set"), QString(), id());
+    QDomElement enable = doc()->createElement(QLatin1String("enable"));
+    enable.setAttribute(QLatin1String("xmlns"), xmlns_carbons);
+    iq.appendChild(enable);
+}
+
+void JT_MessageCarbons::disable()
+{
+    iq = createIQ(doc(), QLatin1String("set"), QString(), id());
+    QDomElement disable = doc()->createElement(QLatin1String("disable"));
+    disable.setAttribute(QLatin1String("xmlns"), xmlns_carbons);
+    iq.appendChild(disable);
+}
+
+void JT_MessageCarbons::onGo()
+{
+    if (!iq.isNull())
+        send(iq);
+}
+
+bool JT_MessageCarbons::take(const QDomElement &e)
+{
+    if (iqVerify(e, Jid(), id())) {
+        if (e.attribute(QLatin1String("type")) != QLatin1String("result"))
```

Comment:
There is no QLatin1String variant for QDN::attribute() so it's waonas you domore allocations.
Use QString::fromLatin1("...")